src/diff/patch.js

```diff
@@ -240,3 +211,108 @@ function patchElement(internal, vnode) {
 		dom._prevValue = newProps.value;
 	}
 }
+
+/**
+ * Render a class component
+ * @param {import('../internal').Internal} internal The component's backing Internal node
+ * @param {import('../internal').VNode} newVNode The new virtual node
+ * @param {any} componentContext The aggregated context for htis component
+ * @returns {import('../internal').ComponentChildren} the component's children
+ */
+function patchComponent(internal, newVNode, componentContext) {
+	let type = /** @type {import('../internal').ComponentType} */ (internal.type);
+	let snapshot,
+		c = internal._component,
+		newProps = newVNode.props;
+
+	if (c._nextState == null) {
+		c._nextState = c.state;
+	}
+
+	if (type.getDerivedStateFromProps != null) {
+		if (c._nextState == c.state) {
+			c._nextState = Object.assign({}, c._nextState);
+		}
+
+		Object.assign(
+			c._nextState,
+			type.getDerivedStateFromProps(newProps, c._nextState)
+		);
+	}
+
+	let oldProps = c.props;
+	let oldState = c.state;
+
+	if (
+		type.getDerivedStateFromProps == null &&
+		newProps !== oldProps &&
+		c.componentWillReceiveProps != null
+	) {
+		c.componentWillReceiveProps(newProps, componentContext);
+	}
+
+	if (
+		(!(internal.flags & FORCE_UPDATE) &&
+			c.shouldComponentUpdate != null &&
+			c.shouldComponentUpdate(newProps, c._nextState, componentContext) ===
+				false) ||
+		(newVNode && newVNode._vnodeId === internal._vnodeId)
+	) {
+		c.props = newProps;
+		c.state = c._nextState;
+		internal.flags |= SKIP_CHILDREN;
+		return;
+	}
+
+	if (c.componentWillUpdate != null) {
+		c.componentWillUpdate(newProps, c._nextState, componentContext);
+	}
+
+	c.context = componentContext;
+	internal.props = c.props = newProps;
+	c.state = c._nextState;
+
+	let renderHook = options._render;
+	if (renderHook) renderHook(internal);
+
+	let counter = 0,
+		renderResult;
+	while (counter++ < 25) {
+		internal.flags &= ~DIRTY_BIT;
+		if (renderHook) renderHook(internal);
+		if (internal.flags & TYPE_CLASS) {
+			renderResult = c.render(c.props, c.state, c.context);
+			// note: disable repeat render invocation for class components
+			break;
+		} else {
+			renderResult = type.call(c, c.props, c.context);
+		}
+		if (!(internal.flags & DIRTY_BIT)) {
+			break;
+		}
+	}
```

Comment:
I had this locally, I believe it's a tiny bit smaller+faster due to the avoidance of a loop for classes.

```suggestion
	// note: disable repeat render invocation for class components
	if (flags & TYPE_CLASS) {
		internal.flags &= ~DIRTY_BIT;
		if (renderHook) renderHook(internal);
		renderResult = inst.render(inst.props, inst.state, inst.context);
	} else {
		let counter = 0;
		while (counter++ < 25) {
			// mark as clean:
			internal.flags &= ~DIRTY_BIT;
			if (renderHook) renderHook(internal);
			renderResult = type.call(inst, inst.props, inst.context);
			// re-render if marked as dirty:
			if (!(internal.flags & DIRTY_BIT)) {
				break;
			}
		}
	}
```